### DISCUSSION TITLE
Some visual tests throw unhandled exceptions?

### DISCUSSION THREAD

**Discussion by hwabis** [ppy/osu#17754] [Q&A] [answered]
Some tests throw one unhandled exception, but I'm able to continue over the line of code and keep running the runner. But when I select the DrawableRoom test (under osu.Game.Tests.Visual.Multiplayer) I get a stack of unhandled exceptions that I can't continue out of, and I have to restart the runner (and quickly select another test when it opens before I get stuck again). Is this expected behavior? I've tried both the Debug and Release configurations.

(I see a bunch of error messages in the debug console.. should I put them here?)

**Comment by peppy** [comment, marked as answer]
A good start would be sharing what you're seeing, yes.

**Reply by hwabis** [reply]
I just realized logs are also created for when running the tester. Here are logs for some tests that throw unhandled exceptions:

Multiplayer DrawableRoom (stops the runner):
[runtime.log](https://github.com/ppy/osu/files/8459201/runtime.log)
[updater.log](https://github.com/ppy/osu/files/8459202/updater.log)
[database.log](https://github.com/ppy/osu/files/8459203/database.log)
[network.log](https://github.com/ppy/osu/files/8459204/network.log)
[performance.log](https://github.com/ppy/osu/files/8459205/performance.log)

 Gameplay SkinEditorMultipleSkins (able to continue):
[updater.log](https://github.com/ppy/osu/files/8459213/updater.log)
[database.log](https://github.com/ppy/osu/files/8459214/database.log)
[performance.log](https://github.com/ppy/osu/files/8459215/performance.log)
[runtime.log](https://github.com/ppy/osu/files/8459216/runtime.log)

**Reply by peppy** [reply]
Thank you, I'll take a look into these.

**Reply by peppy** [reply]
See #17761, #17762 for fixes.

**Reply by hwabis** [reply]
The DrawableRoom test works now but I still have some problems with a few other tests. (Sorry if this is a lot or if I should make separate discussions for each..)

SkinEditorMultipleSkins still throws unhandled exceptions. It stops at line 80 of OsuDifficultyHitObject (`this.lastObject = (OsuHitObject)lastObject;`) with `System.InvalidCastException: 'Unable to cast object of type 'osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider' to type 'osu.Game.Rulesets.Osu.Objects.OsuHitObject'.'` (This also happens with the tests HUDOverlay and SkinnableHUDOverlay).

[updater.log](https://github.com/ppy/osu/files/8466449/updater.log)
[database.log](https://github.com/ppy/osu/files/8466450/database.log)
[performance.log](https://github.com/ppy/osu/files/8466451/performance.log)
[runtime.log](https://github.com/ppy/osu/files/8466452/runtime.log)

If I run the "osu!" step of a test (for example in the Pause tests) and then change to another test while the "osu!" step is still happening then I get `System.OperationCanceledException: 'The operation was canceled.'` at line 110 `cancellationToken.ThrowIfCancellationRequested();` of `DifficultyCalculator`.

[runtime.log](https://github.com/ppy/osu/files/8466502/runtime.log)
[updater.log](https://github.com/ppy/osu/files/8466503/updater.log)
[database.log](https://github.com/ppy/osu/files/8466504/database.log)
[network.log](https://github.com/ppy/osu/files/8466505/network.log)
[performance.log](https://github.com/ppy/osu/files/8466506/performance.log)
 
None of the SkinnableSound tests work for me. They fail at the steps "start sample" and "start sample with looping". (Not an unhandled exception but I thought I'd still put this here)

[runtime.log](https://github.com/ppy/osu/files/8466544/runtime.log)
[updater.log](https://github.com/ppy/osu/files/8466545/updater.log)
[database.log](https://github.com/ppy/osu/files/8466546/database.log)
[performance.log](https://github.com/ppy/osu/files/8466547/performance.log)

**Reply by peppy** [reply]
> The DrawableRoom test works now but I still have some problems with a few other tests. (Sorry if this is a lot or if I should make separate discussions for each..)
> 
> SkinEditorMultipleSkins still throws unhandled exceptions. It stops at line 80 of OsuDifficultyHitObject (`this.lastObject = (OsuHitObject)lastObject;`) with `System.InvalidCastException: 'Unable to cast object of type 'osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider' to type 'osu.Game.Rulesets.Osu.Objects.OsuHitObject'.'` (This also happens with the tests HUDOverlay and SkinnableHUDOverlay).
> 
> [updater.log](https://github.com/ppy/osu/files/8466449/updater.log) [database.log](https://github.com/ppy/osu/files/8466450/database.log) [performance.log](https://github.com/ppy/osu/files/8466451/performance.log) [runtime.log](https://github.com/ppy/osu/files/8466452/runtime.log)

Cannot reproduce. Also there is no error in the attached logs.

> 
> If I run the "osu!" step of a test (for example in the Pause tests) and then change to another test while the "osu!" step is still happening then I get `System.OperationCanceledException: 'The operation was canceled.'` at line 110 `cancellationToken.ThrowIfCancellationRequested();` of `DifficultyCalculator`.
> 
> [runtime.log](https://github.com/ppy/osu/files/8466502/runtime.log) [updater.log](https://github.com/ppy/osu/files/8466503/updater.log) [database.log](https://github.com/ppy/osu/files/8466504/database.log) [network.log](https://github.com/ppy/osu/files/8466505/network.log) [performance.log](https://github.com/ppy/osu/files/8466506/performance.log)
> 

No errors in logs, also can't reproduce. Sounds like your rulesets or build are in a bad state. Please try cleaning things up.

> None of the SkinnableSound tests work for me. They fail at the steps "start sample" and "start sample with looping". (Not an unhandled exception but I thought I'd still put this here)
> 
> [runtime.log](https://github.com/ppy/osu/files/8466544/runtime.log) [updater.log](https://github.com/ppy/osu/files/8466545/updater.log) [database.log](https://github.com/ppy/osu/files/8466546/database.log) [performance.log](https://github.com/ppy/osu/files/8466547/performance.log)

As above.

**Reply by hwabis** [reply]
I just tried deleting the osu folder (under AppData/Roaming/osu on my computer), deleting and re-cloning the osu repo, and uninstalling the lazer binary release, but still no luck. Is there anything else I'm forgetting to try?

**Reply by peppy** [reply]
Could you provide your `client.realm` file and also share how you are running visual tests (ie. what IDE, what command)

**Reply by hwabis** [reply]
[client.zip](https://github.com/ppy/osu/files/8469549/client.zip) (client.realm inside the zip)

I'm using Visual Studio 2022. (I set osu.Game.Tests as the startup project, use the Debug configuration, then press the green play button at the top.)

**Reply by peppy** [reply]
Thanks for the database. I can actually reproduce the third reported issue so will look to fix that. The other two I have no idea about. You'll need to investigate those further if you think they matter. Or just ignore them and move on.

**Reply by hwabis** [reply]
Sounds good, I will let you know if I ever figure those out. I'll just assume they don't matter *too* much since the tests can still pass after passing the breakpoints.
